File: src/pageviews/sub_extract_pageviews.py
#!/usr/bin/env python

import sys
import urllib
import os.path

# the first argument is the name of a text file that contains the title of Wikipedia pages in the list one each line
# the second argument is the file that contains pageviews; the format is "date" "page title" "page view"

if len(sys.argv) < 2:
	print 'usage:', os.path.basename(sys.argv[0]), 'LIST_FILE'
	sys.exit(1)

list_file = sys.argv[1]
if not os.path.exists(list_file):
	print list_file, 'does not exit. failing...'
	sys.exit(1)

titles = []
last_title = ""

for title in open(list_file):
	title = title.strip()
	if title > last_title:
		last_title = title
	titles.append(title)

try:
	for line in sys.stdin:
		fields = line.split()
		if len(fields) > 2:
			title = fields[1]
			if title in titles:
				sys.stdout.write(line)
			elif title > last_title:
				break
except:
	pass
finally:
	sys.stdout.flush()

File: src/pageviews/sub_extract_pageviews.pl
#!/usr/bin/env perl

use Env qw(HOSTNAME);
use URI::Escape;

# check the command-line arguments
#print "$HOSTNAME\$ $0 @ARGV\n";
die "usage: $0 LIST_FILE [DATE]" if (scalar @ARGV < 1 or scalar @ARGV > 2);

%titles = ();
$last_title = "";

open FILE, "<$ARGV[0]" or die "failed opening $ARGV[0]: $!";
while (<FILE>) {
	chomp;
	$titles{uri_unescape($_)} = 1;
}
close FILE;

$date = 'unknown_date';
if ($#ARGV >= 1) {
	$date = $ARGV[1];
}

# main body
%printed = ();
while (<STDIN>) {
	@fields = split;
	if ($#fields >= 1) {
		$title = $fields[1];
		$title =~ s/#.*$//;
		if ($title eq "") {
			next;
		}
		$title = uri_unescape($title);

		if (exists $titles{$title}) {
			$printed{$title} = 1;
			print "$date\t$title\t$fields[2]\t$fields[1]\n";
		}
	}
}

#foreach $title (keys %titles) {
#	if (!$printed{$title}) {
#		print "$date\t$title\t0\n";
#	}
#}
